File: support/kdtree.py
import numpy;

from support import euclidean_distance_sqrt;

class node:
    def __init__(self, data = None, payload = None, left = None, right = None, disc = None, parent = None):
        self.data = data;
        self.payload = payload;
        self.left = left;
        self.right = right;
        self.disc = disc;
        self.parent = parent;
    
    def __repr__(self):
        #return 'Node (%s, %s)' % (self.data, hex(id(self.payload)));
        
        left = None; right = None; parent = None;
        if (self.left is not None):
            left = self.left.data;
            
        if (self.right is not None):
            right = self.right.data;
        
        return 'Node (%s, [%s %s])' % (self.data, left, right);
    
    def __str__(self):
        #return 'Node (%s, %s)' % (self.data, hex(id(self.payload)));
        return self.__repr__();


class kdtree:
    __root = None;
    __dimension = None;
    
    def __init__(self, data_list = None, payload_list = None):
        "Create kd-tree from list of points and from according list of payloads"
        "If lists were not specified then empty kd-tree will be created"
        
        if (data_list is None):
            return; # Just return from here, tree can be filled by insert method later
        
        if (payload_list is None):
            # Case when payload is not specified.
            for index in range(0, len(data_list)):
                self.insert(data_list[index], None);
        else:
            # Case when payload is specified.
            for index in range(0, len(data_list)):
                self.insert(data_list[index], payload_list[index]);
            
                    
    def insert(self, point, payload):
        "Insert new point with payload to kd-tree"
        
        if (self.__root is None):
            self.__dimension = len(point);
            self.__root = node(point, payload, None, None, 0);
            return self.__root;
        
        cur_node = self.__root;
        
        while True:
            if (cur_node.data[cur_node.disc] <= point[cur_node.disc]):
                # If new node is greater or equal than current node then check right leaf
                if (cur_node.right is None):
                    discriminator = cur_node.disc + 1;
                    if (discriminator >= self.__dimension):
                        discriminator = 0;
                        
                    cur_node.right = node(point, payload, None, None, discriminator, cur_node);
                    return cur_node.right;
                else: 
                    cur_node = cur_node.right;
            
            else:
                # If new node is less than current then check left leaf
                if (cur_node.left is None):
                    discriminator = cur_node.disc + 1;
                    if (discriminator >= self.__dimension):
                        discriminator = 0;
                        
                    cur_node.left = node(point, payload, None, None, discriminator, cur_node);
                    return cur_node.left;
                else:
                    cur_node = cur_node.left;
    
    
    def remove(self, point):
        "Remove specified point from kd-tree"
        
        # Get required node
        node_for_remove = self.find_node(point);
        if (node_for_remove is None):
            return None;
        
        parent = node_for_remove.parent;
        node = self.__recursive_remove(node_for_remove);
        if (parent is None):
            self.__root = node;
            
            # If all k-d tree was destroyed
            if (node is not None):
                node.parent = None;
        else:
            if (parent.left is node_for_remove):
                parent.left = node;
            elif (parent.right is node_for_remove):
                parent.right = node;
            else:
                assert 0;   # FATAL ERROR
        
        return self.__root;
    
    
    def __recursive_remove(self, node):
        "Delete node and return root of subtree"
                
        # Check if it is leaf
        if ( (node.right is None) and (node.left is None) ):
            return None;
        
        discriminator = node.disc;
        
        # Check if only left branch exist
        if (node.right is None):
            node.right = node.left;
            node.left = None;
        
        # Find minimal node with in line with coordinate that is defined by discriminator
        minimal_node = self.find_minimal_node(node.right, discriminator);
        parent = self.find_parent(minimal_node);
        
        if (parent.left is minimal_node):
            parent.left = self.__recursive_remove(minimal_node);
        elif (parent.right is minimal_node):
            parent.right = self.__recursive_remove(minimal_node);
        else:
            assert 0;
        
        minimal_node.parent = node.parent;
        minimal_node.disc = node.disc;
        minimal_node.right = node.right;
        minimal_node.left = node.left;
        
        # Update parent for successors of previous parent.
        if (minimal_node.right is not None):
            minimal_node.right.parent = minimal_node;
             
        if (minimal_node.left is not None):
            minimal_node.left.parent = minimal_node;
        
        return minimal_node;
        
    
    def find_minimal_node(self, node, discriminator):
        "Return node minimal coordinate that is defined by discriminator"
        assert node is not None;
        candidates = [self.find_minimal_node(child, discriminator) for child in self.children(node) if child is not None];
        candidates = candidates + [ node ];
        
        min_key = lambda cur_node: cur_node.data[discriminator];
        return min(candidates, key = min_key);
    
    
    def find_parent(self, node, cur_node = None):
        "Return parent of specified node"
        "The parent was added for reducing execution time, so the method is used for compatibility, so should be removed"
        # TODO: refactoring is required for removing this method.
        return node.parent;
    
    
    def find_node(self, point, cur_node = None):
        "Return node with coordinates that are defined by specified point"
        "If node does not exist then None will be returned"
        "Otherwise required node will be returned"
        
        req_node = None;
        
        if (cur_node is None):
            cur_node = self.__root;
        
        while True:
            if (cur_node.data[cur_node.disc] <= point[cur_node.disc]):
                # Check if it's required node
                if (cur_node.data == point):
                    req_node = cur_node;
                    break;
                
                if (cur_node.right is not None):
                    cur_node = cur_node.right;
                else:
                    break;  # assert 0
            
            else:
                if (cur_node.left is not None):
                    cur_node = cur_node.left;
                else:
                    break;  # assert 0
                
        return req_node;
    
    
    
    def find_nearest_dist_node(self, point, distance, retdistance = False):
        "Return nearest neighbor in area with radius = distance"
        best_nodes = self.find_nearest_dist_nodes(point, distance);
            
        if (best_nodes == []): 
            return None;
        
        nearest = min(best_nodes, key = lambda item: item[0]);
        
        if (retdistance == True):
            return nearest;
        else:
            return nearest[1];
    
    
    def find_nearest_dist_nodes(self, point, distance):
        "Return list of neighbors such as tuple (distance, node) that is located in area that is covered by distance"
        best_nodes = [];
        self.__recursive_nearest_nodes(point, distance, distance ** 2, self.__root, best_nodes);
        
        return best_nodes;
    
    
    def __recursive_nearest_nodes(self, point, distance, sqrt_distance, node, best_nodes):
        minimum = node.data[node.disc] - distance;
        maximum = node.data[node.disc] + distance;
        
        if (node.right is not None):
            if (point[node.disc] >= minimum):
                self.__recursive_nearest_nodes(point, distance, sqrt_distance, node.right, best_nodes);
        
        if (node.left is not None):
            if (point[node.disc] < maximum):
                self.__recursive_nearest_nodes(point, distance, sqrt_distance, node.left, best_nodes);
        
        candidate_distance = euclidean_distance_sqrt(point, node.data);
        if (candidate_distance <= sqrt_distance):
            best_nodes.append( (candidate_distance, node) );
    
    
    # TODO: This method is not ready yet
    def find_nearest_node(self, point):
        "Search the nearest node of the given point"
        "Note: Does not work properly"
        cur_node = self.__root;
        
        best_node = None;
        best_distance = numpy.Inf;
        
        while True:
            # Check if it's best candidate and maybe it's owner of the coordinates.
            candidate_distance = euclidean_distance_sqrt(cur_node.data, point);
            if ((candidate_distance < best_distance) and (candidate_distance != 0)):
                best_node = cur_node;
                best_distance = candidate_distance;

            # Sort the children, nearer one first
            children = iter( sorted(self.children(cur_node), key = lambda node: euclidean_distance_sqrt(node.data[cur_node.disc], point[cur_node.disc])) );

            c1 = next(children, None);
            if c1:
                cur_node = c1;
                continue;

            c2 = next(children, None);
            if c2 and ( euclidean_distance_sqrt(cur_node.data[cur_node.disc], point[cur_node.disc]) < best_distance ):
                cur_node = c2;
                continue;

            return best_node;
    
    
    def children(self, node):
        if (node.left is not None):
            yield node.left;
        if (node.right is not None):
            yield node.right;
            
    
    def traverse(self, node = None, level = None):
        if (node is None):
            node  = self.__root;
            level = 0;
        
        if (node is None):
            return [];
        
        items = [ (level, node) ];        
        for child in self.children(node):
            if child is not None:
                items += self.traverse(child, level + 1);
        
        return items;
            
    
    def show(self):
        nodes = self.traverse();
        if (nodes == []):
            return;
        
        nodes.sort(key = lambda item: item[0]);
        
        level = nodes[0];
        string = "";
        for item in nodes:
            if (level == item[0]):
                string += str(item[1]) + "\t";
                
            else:
                print(string);
                level = item[0];
                string = str(item[1]) + "\t";
                
        print(string);
    

File: core/ccore/ccore/kdtree.cpp
#include "kdtree.h"
#include "support.h"

kdnode::kdnode(std::vector<double> * p_data, void * p_payload,  kdnode * p_left, kdnode * p_right, kdnode * p_parent, unsigned int disc) :
	data(p_data), payload(p_payload), left(p_left), right(p_right), parent(p_parent), discriminator(disc)
	{ }

kdnode::~kdnode() { }


/***********************************************************************************************
*
* @brief   Default constructor.
*
***********************************************************************************************/
kdtree::kdtree() : root(NULL), dimension(0) { }

/***********************************************************************************************
*
* @brief   Constructor of kd tree with pre-defined information.
*
* @param   data               - coordinates that describe nodes in tree.
* @param   payloads           - payloads of nodes (can be NULL if it's not required).
*
***********************************************************************************************/
kdtree::kdtree(const std::vector< std::vector<double> *> * data, const std::vector<void *> * payloads) {
	root = NULL;
	dimension = (*(*data)[0]).size();

	if (payloads) {
		if (data->size() != payloads->size()) {
			throw std::runtime_error("Number of points should be equal to number of according payloads");
		}

		for (unsigned int index = 0; index < data->size(); index++) {
			insert((*data)[index], (*payloads)[index]);
		}
	}
	else {
		for (unsigned int index = 0; index < data->size(); index++) {
			insert((*data)[index], NULL);
		}
	}
}

/***********************************************************************************************
*
* @brief   Default destructor.
*
***********************************************************************************************/
kdtree::~kdtree(void) {
	/* traverse tree and remove each node */
}

/***********************************************************************************************
*
* @brief   Insert new node in the tree.
*
* @param   point              - coordinates that describe node in tree.
* @param   payload            - payloads of node (can be NULL if it's not required).
*
* @return  Pointer to added node in the tree.
*
***********************************************************************************************/
kdnode * kdtree::insert(std::vector<double> * point, void * payload) {
	if (root == NULL) {
		kdnode * node = new kdnode(point, payload, NULL, NULL, NULL, 0);
		root = node;
		return node;
	}
	else {
		kdnode * cur_node = root;

		while(true) {
			/* If new node is greater or equal than current node then check right leaf */
			if (*cur_node <= *point) {
				if (cur_node->get_right() == NULL) {
					unsigned int discriminator = cur_node->get_discriminator() + 1;
					if (discriminator >= dimension) {
						discriminator = 0;
					}

					cur_node->set_right(new kdnode(point, payload, NULL, NULL, cur_node, discriminator));
					return cur_node->get_right();
				}
				else {
					cur_node = cur_node->get_right();
				}
			}
			/* If new node is less than current then check left leaf */
			else {
				if (cur_node->get_left() == NULL) {
					unsigned int discriminator = cur_node->get_discriminator() + 1;
					if (discriminator >= dimension) {
						discriminator = 0;
					}

					cur_node->set_left(new kdnode(point, payload, NULL, NULL, cur_node, discriminator));
					return cur_node->get_left();
				}
				else {
					cur_node = cur_node->get_left();
				}
			}
		}
	}
}

/***********************************************************************************************
*
* @brief   Remove point with specified coordinates.
*
* @param   point              - coordinates that describe node in tree.
*
***********************************************************************************************/
void kdtree::remove(std::vector<double> * point) {
	kdnode * node_for_remove = find_node(point);
	if (node_for_remove == NULL) {
		return;
	}

	remove(node_for_remove);
}

/***********************************************************************************************
*
* @brief   Remove node from the tree.
*
* @param   node_for_remove    - pointer to node that is located in tree.
*
***********************************************************************************************/
void kdtree::remove(kdnode * node_for_remove) {
	kdnode * parent = node_for_remove->get_parent();
	kdnode * node = recursive_remove(node_for_remove);

	if (parent == NULL) {
		root = node;

		/* if tree is almost destroyed */
		if (node != NULL) {
			node->set_parent(NULL);
		}
	}
	else {
		if (parent->get_left() == node_for_remove) {
			parent->set_left(node);
		}
		else if (parent->get_right() == node_for_remove) {
			parent->set_right(node);
		}
		else {
			throw std::runtime_error("Structure of KD Tree is corrupted");
		}
	}
}

/***********************************************************************************************
*
* @brief   Recursive remove of node in tree.
*
* @param   node            - node that should be removed.
*
* @return  Node that should replace removed node (if it's not leaf).
*
***********************************************************************************************/
kdnode * kdtree::recursive_remove(kdnode * node) {
	/* Check if it's a leaf */
	if ( (node->get_right() == NULL) && (node->get_left() == NULL) ) {
		return NULL;
	}

	unsigned int discriminator = node->get_discriminator();

	/* Check if only left branch exist */
	if (node->get_right() == NULL) {
		node->set_right(node->get_left());
		node->set_left(NULL);
	}

	/* Find minimal node in line with coordinate that is defined by discriminator */
	kdnode * minimal_node = find_minimal_node(node->get_right(), discriminator);
	kdnode * parent = minimal_node->get_parent();

	if (parent->get_left() == minimal_node) {
		parent->set_left(recursive_remove(minimal_node));
	}
	else if (parent->get_right() == minimal_node) {
		parent->set_left(recursive_remove(minimal_node));
	}
	else {
		throw std::runtime_error("Structure of KD Tree is corrupted");
	}

	minimal_node->set_parent(node->get_parent());
	minimal_node->set_discriminator(node->get_discriminator());
	minimal_node->set_right(node->get_right());
	minimal_node->set_left(node->get_left());

	/* Update parent for successors of previous parent */
	if (minimal_node->get_right() != NULL) {
		minimal_node->get_right()->set_parent(minimal_node);
	}

	if (minimal_node->get_left() != NULL) {
		minimal_node->get_left()->set_parent(minimal_node);
	}

	return minimal_node;
}

/***********************************************************************************************
*
* @brief   Find minimal node in subtree in line with specified discriminator.
*
* @param   node            - root of subtree where searching should be performed.
* @param   discriminator   - discriminator that is used for comparison of nodes.
*
* @return  Return the smallest node in specified subtree in line with discriminator.
*
***********************************************************************************************/
kdnode * kdtree::find_minimal_node(kdnode * node, unsigned int discriminator) {
	kdnode * minimal_node = node;

	std::vector<kdnode *> * children = node->get_children();
	for (unsigned int index_children = 0; index_children < children->size(); index_children++) {
		kdnode * candidate = find_minimal_node((*children)[index_children], discriminator);

		if (candidate->get_value(discriminator) < minimal_node->get_value(discriminator)) {
			minimal_node = candidate;
		}
	}

	delete children;
	children = NULL;

	return minimal_node;
}

/***********************************************************************************************
*
* @brief   Find node in tree using coordinates.
*
* @param   point              - coordinates of searched node.
*
* @return  Pointer to found node in tree.
*
***********************************************************************************************/
kdnode * kdtree::find_node(std::vector<double> * point) {
	return find_node(point, root);
}

/***********************************************************************************************
*
* @brief   Find node in tree using coordinates in subtree.
*
* @param   point              - coordinates of searched node.
* @param   cur_node           - root of subtree.
*
* @return  Pointer to found node in tree.
*
***********************************************************************************************/
kdnode * kdtree::find_node(std::vector<double> * point, kdnode * node) {
	kdnode * req_node = NULL;
	kdnode * cur_node = node;

	while(true) {
		if (*cur_node <= *point) {
			if (*cur_node == *point) {
				req_node = cur_node;
				break;
			}

			if (cur_node->get_right() != NULL) {
				cur_node = cur_node->get_right();
			}
			else {
				throw std::runtime_error("Structure of KD Tree is corrupted");
			}
		}
		else {
			if (cur_node->get_left() != NULL) {
				cur_node = cur_node->get_left();
			}
			else {
				throw std::runtime_error("Structure of KD Tree is corrupted");
			}
		}
	}

	return req_node;
}


/***********************************************************************************************
*
* @brief   Default constructor. Search will not be performed until it's initialized.
*
***********************************************************************************************/
kdtree_searcher::kdtree_searcher() : distance(0), sqrt_distance(0), initial_node(NULL), search_point(NULL) { }

/***********************************************************************************************
*
* @brief   Default destructor.
*
***********************************************************************************************/
kdtree_searcher::~kdtree_searcher() {
	if (nodes_distance != NULL) {
		delete nodes_distance;
		nodes_distance = NULL;
	}

	if (nearest_nodes != NULL) {
		delete nearest_nodes;
		nearest_nodes = NULL;
	}
}

/***********************************************************************************************
*
* @brief   Constructor of searcher with request for searching.
*
* @param   (in) point              - point for which nearest nodes should be found.
* @param   (in) node               - initial node in tree from which searching should started.
* @param   (in) radius_search      - allowable distance for searching from the point.
*
***********************************************************************************************/
kdtree_searcher::kdtree_searcher(std::vector<double> * point, kdnode * node, const double radius_search) {
	initialize(point, node, radius_search);
}

/***********************************************************************************************
*
* @brief   Initialization of new request for searching.
*
* @param   (in) point              - point for which nearest nodes should be found.
* @param   (in) node               - initial node in tree from which searching should started.
* @param   (in) radius_search      - allowable distance for searching from the point.
*
***********************************************************************************************/
void kdtree_searcher::initialize(std::vector<double> * point, kdnode * node, const double radius_search) {
	distance = radius_search;
	sqrt_distance = radius_search * radius_search;

	initial_node = node;
	search_point = point;
}

/***********************************************************************************************
*
* @brief   Prepare storages that are required for searching.
*
***********************************************************************************************/
void kdtree_searcher::prepare_storages() {
	nodes_distance = new std::vector<double>();
	nearest_nodes = new std::vector<kdnode *>();
}

/***********************************************************************************************
*
* @brief   Recursive method for searching nodes that satisfy the request.
*
* @param   node               - initial node in tree from which searching should performed.
*
***********************************************************************************************/
void kdtree_searcher::recursive_nearest_nodes(kdnode * node) {
	double minimum = node->get_value() - distance;
	double maximum = node->get_value() + distance;

	if (node->get_right() != NULL) {
		if ((*search_point)[node->get_discriminator()] >= minimum) {
			recursive_nearest_nodes(node->get_right());
		}
	}

	if (node->get_left() != NULL) {
		if ((*search_point)[node->get_discriminator()] < maximum) {
			recursive_nearest_nodes(node->get_left());
		}
	}

	double candidate_distance = euclidean_distance_sqrt(search_point, node->get_data());
	if (candidate_distance <= sqrt_distance) {
		nearest_nodes->push_back(node);
		nodes_distance->push_back(candidate_distance);
	}
}

/***********************************************************************************************
*
* @brief   Search nodes that are located in specified distance from specified point.
*
* @return  Return vector of found nodes in kd tree that satisfy the request.
*
***********************************************************************************************/
std::vector<kdnode *> * kdtree_searcher::find_nearest_nodes(std::vector<double> * distances) {
	prepare_storages();

	std::vector<kdnode *> * result = nearest_nodes;

	recursive_nearest_nodes(initial_node);

	nearest_nodes = NULL; /* application responds for the vector */

	if (distances != NULL) {
		distances = nodes_distance; /* application responds for the vector */
		nodes_distance = NULL;
	}
	else {
		delete nodes_distance;
		nodes_distance = NULL;
	}

	return result;
}

/***********************************************************************************************
*
* @brief   Search the nearest node in specified location for specified point in the request.
*
* @return  Return pointer to the nearest node in kd tree that satisfy the request.
*
***********************************************************************************************/
kdnode * kdtree_searcher::find_nearest_node() {
	prepare_storages();

	kdnode * node = NULL;

	recursive_nearest_nodes(initial_node);

	if (nodes_distance->size() > 0) {
		double minimal_distance = (*nodes_distance)[0];
		for (unsigned int index = 1; index < nodes_distance->size(); index++) {
			if ((*nodes_distance)[index] < minimal_distance) {
				minimal_distance = (*nodes_distance)[index];
				node = (*nearest_nodes)[index];
			}
		}
	}

	delete nearest_nodes;
	nearest_nodes = NULL;

	delete nodes_distance;
	nodes_distance = NULL;

	return node;
}
